### ISSUE TITLE
Proper node relationships throughout Flyte

### ISSUE THREAD

**Issue by wild-endeavor** [flyteorg/flyte#218]
**Motivation: Why do you think this is important?**
The graph relationship in Flyte workflows today depends on task executions.  This is different than a traditional graph where there's just nodes and they point to each other.

For instance, let's look at 
<img width="641" alt="Screen Shot 2020-03-20 at 7 31 26 PM" src="https://user-images.githubusercontent.com/2896568/77217801-681b6380-6ae2-11ea-8a2d-08c6e59e95f4.png">

This is a dynamic task execution and it clearly shows that node "fnzejraq" belongs under the "task" node.  How is this done?

Admin has a table called `node_executions` that has a flat list of all the node executions done for this workflow run.  If you look up the record for `node_id = 'fnzejraq'` you will see that it has a `parent_task_execution_id` filled in, let's say it's `123`.

In the `task_executions` table, for record `123`, the `node_id` column says `task1`.  This is how Admin keeps track of the relationships between nodes.  An execution of a task acts as the edge in the graph.

There are reasons why it was done this way.  (@EngHabu - can you please help elaborate the reasons?), but issues around how to handle multiple instantiations of subworkflows (and hence duplicate subworkflow node IDs) is one and handling around retries is another.

**Goal: What should the final outcome look like, ideally?**
I would like to see proper node to node relationships with proper handling of the above issues.  Using task executions makes proper display of subworkflows a bit of a hack (basically we have to invent a mock task execution), and I think given the graphy nature of Flyte, we should have a proper data model for a graph.

**Describe alternatives you've considered**
None, not set on having this, but I think this is something we should have.

**Flyte component**
- [ ] Overall
- [ ] Flyte Setup and Installation scripts
- [ ] Flyte Documentation
- [ ] Flyte communication (slack/email etc)
- [x] FlytePropeller
- [x] FlyteIDL (Flyte specification language)
- [ ] Flytekit (Python SDK)
- [x] FlyteAdmin (Control Plane service)
- [ ] FlytePlugins
- [ ] DataCatalog 
- [ ] FlyteStdlib (common libraries)
- [x] FlyteConsole (UI)
- [ ] Other

**[Optional] Propose: Link/Inline**

**Additional context**

**Is this a blocker for you to adopt Flyte**
No.

**Comment by EngHabu** [comment]
RE: Reasons for what we have,
A few assumptions first:
- Node IDs (strings) are assumed to be unique within a given workflow
- A node controls retries of its target (e.g. a task) but in of itself doesn't retry.

The current model allows this model representation:
n1:
  -> task1 retry 0:
       -> fabc21 (hash value of the lineage of this node + node ID to ensure uniqueness within the entire graph)
        -> fcd13
  -> task1 retry 1:
       -> fabx21 (hashes are different because they account for the retry attempt from the task)
       -> fcdx34 

It's important for to maintain the ability for the UI to show this relationship as presented above.

However, this causes a few undesired side-effects

1. Those hashes for node IDs obfuscate the real node IDs yielded by the task making it hard to distinguish between them

2. Node -> Node relationship are not allowed.

As for a proposal to how to represent this:
I think we should fundamentally change the UX representation (And I would like to start the conversation with UX not with the data model since that's super fluid)
n1:
   -> Attempt 1 (or 0)
   -> Attempt 2

1) If the node is a TaskNode (e.g. a Python Task), each Attempt should be a link, clicking that opens the side pane with task execution details... etc.

2) If the node is a Workflow node (with a launchPlan or a subworkflow), each Attempt should just have the arrow beside it to see the sub-nodes

3) If the node is a Dynamic Node, each Attempt should just have an arrow beside it to see the generator task as well as all the nodes that got generated... The task should be a link that opens up the side pane to view the task information as well as the longs links... etc.

All sub list should be ordered by start time.

If this is agreed upon, then to model that, we will have to change how the node table key is structured
node_executions columns:
Primary Key [execID, nodeID, attempt, lineageHash]
parentNodeID,
parentAttempt
parentLineageHash
...

The reason linageHash is required is to allow us to store the original NodeIDs without fear of collision. 

task_executions columns:
Primary Key [execID, taskID, nodeID, attempt, lineageHash]

A similar change will need to happen for the launch plan executions table

This will allow us to store tasks with nodes as parents as well as nodes with nodes as parents.
We should then have a new endpoint of /sub_executions/<node primary key>
to retrieve a list of 
```
message SubExecution {
    oneof Target {
        TaskExecution task_execution = 1;
        NodeExecution node_execution = 2;
        LaunchPlanExecution workflow_execution = 3;
    }
}
```

What do you think?

CC @katrogan @kumare3 too

**Comment by schottra** [comment]
@EngHabu @wild-endeavor

Regarding the UI showing the relationship: 
I'm already in the process of changing to an experience that is very similar to what you've mentioned. The important detail is that we no longer want to show the relationship as Node -> Task -> Node, but as Node -> Node, with the task executions / attempts being explorable details of each node.

Right now, the data model is blocking this from being an efficient query. It's not possible for me to determine the following things in a single request:
* Does a Node Execution have children that were generated at runtime and thus are not discoverable in the graph structure?
* What are the Node Executions which are direct children of a given Node Execution?

Both of these require cascading network requests that scale quite poorly (N * M) in relation to the number of Node Executions in a given list request.

So, I'm very much in favor of restructuring the data to allow for Node -> Node relationships.

**Comment by kumare3** [comment]
@anandswaminathan I think this is done right, can we close this out?

**Comment by anandswaminathan** [comment]
@kumare3 Yes. We need to add new fields here https://github.com/flyteorg/flytekit/blob/master/flytekit/models/node_execution.py

Will close this once I add that.

**Comment by kumare3** [comment]
@anandswaminathan Is this done? can we please close this out. Node-node relationships are now default

**Comment by anandswaminathan** [comment]
@kumare3 Wait. I am sending Flytekit PR and will close

**Comment by kumare3** [comment]
@anandswaminathan is this ready to be closed?

**Comment by wild-endeavor** [comment]
@anandswaminathan is this ready to be closed?

**Comment by wild-endeavor** [comment]
Can you tell me what needs to be done in flytekit and i'll make a separate ticket for flytekit.

**Comment by kumare3** [comment]
@anandswaminathan have you done this? This is mostly implemented - can you close this out?

**Comment by kumare3** [comment]
The flytekit change is trivial and just needs additonal fields - https://github.com/flyteorg/flyte/issues/555. Closing the issue
